lib/kafka_ex/partitioner.ex

```diff
@@ -0,0 +1,43 @@
+defmodule KafkaEx.Partitioner do
+  @moduledoc """
+  Behaviour definition for partitioners, that assigns partitions for requests.
+  """
+
+  alias KafkaEx.Protocol.Produce.Request, as: ProduceRequest
+  alias KafkaEx.Protocol.Metadata.Response, as: MetadataResponse
+
+  @callback assign_partition(request :: %ProduceRequest{}, metadata :: %MetadataResponse{}) ::
+    %ProduceRequest{}
+
+  defmacro __using__(_) do
```

Comment:
Any reason why this is a macro?

Reply:
Isn't it a common pattern to make behaviour defining modules define `__using__` macro, so any module that would implement the behaviour could `use BehaviousModule`?

I made `__using__` macro define `get_key` function as I believe it would be a common function used by any custom partitioner implementation.

Reply:
I think we could just make `get_key/1` a function on this module, which could be imported or used elsewhere if needed. I generally like to shy away from defining functions on another module if not needed.